scvi/inference/cite.py

```diff
@@ -0,0 +1,215 @@
+            # general fish case
+            sample_batch, local_l_mean, local_l_var, batch_index, labels = tensors[:5]
+            # For loop could probably be removed through n_samples parameter
+            for i in range(mc_samples):
+                means = vae.get_sample_rate(sample_batch, mode='adt')
+                var_disp = vae.get_sample_dispersion(sample_batch, mode='adt')
+
+                # Parametrized as r, p
+                if vae.reconstruction_loss_adt == 'nb':
+                    r = var_disp
+                    p = means / (means + var_disp)
+                    odds = p / (1 - p)
+                    rate = Gamma(concentration=r, rate=1 / odds).sample()
+                    if i == 0:
+                        # samples = NegativeBinomial(total_count=var_disp, probs=means / (means + var_disp)).sample()
+                        samples = torch.poisson(rate)
+                    else:
+                        # samples += NegativeBinomial(total_count=var_disp, probs=means / (means + var_disp)).sample()
+                        samples += torch.poisson(rate)
+                elif vae.reconstruction_loss_adt == 'poisson':
+                    if i == 0:
+                        samples = Poisson(means).sample()
+                    else:
+                        samples += Poisson(means).sample()
+                # log_normal
+                else:
+                    if i == 0:
+                        samples = LogNormal(means, var_disp).sample()
+                    else:
+                        samples += LogNormal(means, var_disp).sample()
+
+            # Take average sample
+            samples /= mc_samples
+
+            umi_ = sample_batch[:, :vae.n_input_genes]
+            reconst_loss_umi, reconst_loss_adt, kl_divergence = vae(torch.cat((umi_, samples), 1), local_l_mean,
+                                                                    local_l_var, batch_index=batch_index,
+                                                                    y=labels, **kwargs)
+            log_lkl += torch.sum(reconst_loss_adt).item()
+
+            adt_ = sample_batch[:, vae.n_input_genes:]
+            error = torch.sum((samples - adt_)**2, dim=0)
+            mse += error
+
+            mae += torch.sum(torch.abs(samples - adt_), dim=0)
+
+        n_samples = len(self.indices)
+        mse /= n_samples
+        mae /= n_samples
+        log_lkl /= n_samples
+        return mse, mae, log_lkl
+
+    def differential_expression_stats(self, M_sampling=100, mode='adt'):
+        r"""
+        Output average over statistics in a symmetric way (a against b)
+        forget the sets if permutation is True
+        :param vae: The generative vae and encoder network
+        :param data_loader: a data loader for a particular dataset
+        :param M_sampling: number of samples
+        :param mode: UMI or ADT data
+        :return: A 1-d vector of statistics of size n_genes
+        """
+        px_scales = []
```

Comment:
This appears to be duplicated code, found in another file. It's best to keep blocks of duplicated code (say, more than 1 or 2 lines) out of the master branch. Otherwise the code base gets unwieldy pretty fast.